app/merchant/views.py

```diff
@@ -87,46 +70,53 @@ def listing_info(listing_id):
     abort(404)
 
 
-@merchant.route('/add_to_cart')
+@merchant.route('/add_to_cart/<int:listing_id>', methods=["PUT"])
 @login_required
 @merchant_required
-def add_to_cart():
-    current_listing_id = request.args.get('current_listing_id')
-    quantity_needed = request.args.get('quantity_needed')
-    if not current_listing_id or not quantity_needed:
-        return redirect(url_for('.index'))
-    listing = Listing.query.filter_by(id=current_listing_id).first()
-    cart_item = CartItem.query.filter_by(
-        merchant_id=current_user.id).filter_by(
-        listing_id=current_listing_id).first()
-
-    if listing is None:
+def add_to_cart(listing_id):
+    """Adds listing to cart with specified quantity"""
+    listing = Listing.query.filter_by(id=listing_id, available=True).first()
+    if not listing:
         abort(404)
-
-    if cart_item is None:
-        db.session.add(CartItem(
-            merchant_id=current_user.id,
-            listing_id=current_listing_id,
-            quantity=quantity_needed
-        ))
-    else:
+    if not request.json:
+        abort(400)
+    if ('quantity' in request.json and
+                type(request.json['quantity']) is not int):
+        abort(400)
+    cart_item = CartItem.query.filter_by(merchant_id=current_user.id,
+                                         listing_id=listing_id).first()
+    new_quantity = request.json['quantity']
+    is_currently_incart = cart_item is not None
+
+    if new_quantity == 0 and is_currently_incart:
         db.session.delete(cart_item)
-
+    elif new_quantity != 0 and is_currently_incart:
+        cart_item.quantity = new_quantity
+    elif new_quantity != 0 and not is_currently_incart:
+        db.session.add(CartItem(merchant_id=current_user.id,
+                                listing_id=listing_id,
+                                quantity=new_quantity))
     db.session.commit()
-    return redirect(url_for('.listing_view_all'))
+    return jsonify({'quantity': new_quantity})
 
 
-@merchant.route('/add_to_favorites')
+@merchant.route('/change_favorite/<int:listing_id>', methods=['PUT'])
 @login_required
 @merchant_required
-def favorite():
-    listing_id = request.args.get('listing_id')
-    if not listing_id:
+def change_favorite(listing_id):
+    listing = Listing.query.filter_by(id=listing_id).first()
+    if not listing:
         abort(404)
-    listing = Listing.query.filter_by(id=listing_id).first_or_404()
-    if listing in current_user.bookmarks:
-        current_user.bookmarks.remove(listing)
-    else:
+    if not request.json:
+        abort(400)
+    if ('isFavorite' in request.json and
+                type(request.json['isFavorite']) is not bool):
```

Comment:
if ('isFavorite' not in request.json or type(request.json['isFavorite']) is not bool):